### DISCUSSION TITLE
POC for a Hardhat project!!

### DISCUSSION THREAD

**Discussion by 0xnayan** [Cyfrin/security-and-auditing-full-course-s23#57] [Q&A] [answered]
For competitive audit, if the protocol is using hardhat and test is written in javascript now for Proof of Code can I submit a test using foundry or I have to do that in javascript??

**Comment by johnsonstephan** [comment, marked as answer]
Hi @0xnayan, you raise an important question! This tends to be a puzzling area for some auditors. 

**TL;DR: The answer is yes, you can.** 

## Let’s get into more detail
In a competitive audit, if a protocol uses Hardhat with JavaScript tests, you may in fact submit a PoC using Foundry. 

For example, this approach was successfully taken in the competitive audit for ENS on Code4rena (C4). 

- In the C4 ENS audit report, a PoC for [finding M-01](https://code4rena.com/reports/2023-10-ens#m-01-some-tokens-enable-the-direct-draining-of-all-approved-erc20votes-tokens) was submitted and credited using Foundry -- even though ENS uses Hardhat, as noted in their [C4 ENS repo](https://github.com/code-423n4/2023-10-ens).
- The audit report even highlights a "practical step" of converting the Hardhat tests into a Foundry environment in [Section 2.2](https://code4rena.com/reports/2023-10-ens#2-audit-approach).

> [!TIP]
> However, it's important to review the guidelines of any audits you participate in, to ensure there are no restrictions on your choice of framework.

<details><summary>Bonus Information</summary>
<p>

1. Considerations: When deciding between the two, a key factor is that Foundry allows you to write your tests in Solidity, which will usually align with the development language of the protocol. Other advantages of Foundry include faster test and script execution, native fuzz testing capabilities, and “cheat codes” for state manipulation.
2. Resource: Additionally, you may find @PatrickAlphaC's [hilarious 1-minute clip](https://www.youtube.com/watch?v=j_bA2Ykm7h0) on using Foundry useful (and extremely entertaining). 

</p>
</details> 

Does all of that make sense?

**Reply by 0xnayan** [reply]
Thank you @johnsonstephan 🙏

**Reply by johnsonstephan** [reply]
@0xnayan No problem, happy to help!
